File: 151.翻转字符串里的单词.py
#
# @lc app=leetcode.cn id=151 lang=python
#
# [151] 翻转字符串里的单词
#

# @lc code=start
class Solution(object):
    def reverseWords(self, s):
        """
        :type s: str
        :rtype: str
        """
        # strip spaces
        if not s: return s
        s = list(s)
        sp, fp = 0, 0
        in_word = False
        while fp < len(s):
            if s[fp] != ' ':
                if not in_word: in_word = True
                s[sp], s[fp] = s[fp], s[sp]
                sp += 1
            else:
                if in_word: 
                    in_word = False
                    sp += 1
            fp += 1
        # cut the tail of spaces
        fp = len(s) - 1
        while fp >= 0 and s[fp] == ' ':
            fp -= 1
        s = s[:fp + 1]
        # reverse the char array
        self.reverse_str(0, len(s) - 1, s)
        
        # reverse the words
        sp, fp = 0, 0
        while fp < len(s):
            while fp < len(s) and s[fp] != ' ':
                fp += 1
            self.reverse_str(sp, fp - 1, s)
            fp += 1
            sp = fp

        return ''.join(s)
    
    def reverse_str(self, sp, fp, s):
        while sp < fp:
            s[sp], s[fp] = s[fp], s[sp]
            sp += 1
            fp -= 1

                
# @lc code=end

File: 151.翻转字符串里的单词.c
/*
 * @lc app=leetcode.cn id=151 lang=c
 *
 * [151] 翻转字符串里的单词
 */

// @lc code=start


char * reverseWords(char * s){
    // strip
    int i, j, k, m;
    for (i = strlen(s) - 1; i >= 0; i--)
    {
        if (*(s + i) != ' ')
        {
            *(s + i + 1) = '\0';
            break;
        }
    }
    if (i == -1 && *(s + i + 1) != '\0')
        *s = '\0';
    for (i = 0; i < strlen(s); i++)
    {
        if (*(s + i) != ' ')
        {
            s += i;
            break;
        }
    }
    bool inword = true;
    for (i = 0, j = 0; *(s + j) != '\0'; )
    {
        if (*(s + j) == ' ')
        {
            if (inword)
            {
                *(s + i++) = *(s + j++);
                inword = false;
            }
            else
                j++;
        }
        else
        {
            *(s + i++) = *(s + j++);
            if (!inword)
                inword = true;
        }
    }
    *(s + i) = '\0';
    for (i = 0, j = strlen(s) - 1; i < j; i++, j--)
    {
        char tmp = *(s + i);
        *(s + i) = *(s + j);
        *(s + j) = tmp;
    }
    for (j = 0, k = 0; j < strlen(s) + 1; j++)
    {
        if (*(s + j) == ' ' || *(s + j) == '\0')
        {
            for (i = k, m = j - 1; i < m; i++, m--)
            {
                char tmp = *(s + i);
                *(s + i) = *(s + m);
                *(s + m) = tmp;
            }
            k = j + 1;
        }
    }
    return s;
}


// @lc code=end

